synapse/handlers/devicemessage.py

```diff
@@ -135,12 +142,16 @@ async def on_direct_to_device_edu(self, origin: str, content: JsonDict) -> None:
                 message_type, sender_user_id, by_device
             )
 
-        stream_id = await self.store.add_messages_from_remote_to_device_inbox(
+        # Add messages to the database.
+        # Retrieve the stream id of the last-processed to-device message.
+        max_stream_id = await self.store.add_messages_from_remote_to_device_inbox(
```

Comment:
What does the `max` indicate here? Largest stream id that has been processed?

Reply:
Yes, it's just the stream id of the last to-device message that was added to the database. Messages are added to the database in with incrementing stream ids.

```
# Retrieve the stream id of the to-device message that was stored last (the largest stream id).
```

instead perhaps?

Reply:
I guess my concern here was that "processing" could be a separate step to persisting them in the database. I also wasn't sure if we might be adding messages with stream ids say `10, 11, 12`, but someone else has already put a message with a stream id of `30` into the db.

Reply:
The stream id that is assigned to each to-device message is implemented using an ever-incrementing database sequence:

https://github.com/matrix-org/synapse/blob/4535532526581834ab798996ffe73f6d19c25123/synapse/storage/databases/main/deviceinbox.py#L468-L475

So having multiple rows with the same stream id should never occur.

to-device messages are stored here. Later on they are retrieved from the database and sent off to remote homeservers/down user sync streams. Each remote has a last-known-received stream id, and the same is encoded in client sync tokens.

...does that quell your concern?

Reply:
`last_stream_id` might be a better name in this case?